### PR TITLE
Image refactor

### PR DESCRIPTION
**Pull request by Ghabry**
This ends what @carstene1ns started :+1:

### CODE REVIEW

File: src/bitmap.cpp [hunk 1]
```diff
@@ -533,19 +551,38 @@ Bitmap::Bitmap(const std::string& filename, bool transparent, uint32_t flags) {
...
+		free(pixels);
+
+		pixels = nullptr;
+
+		return;
+	}
+
 	Init(w, h, (void *) NULL);
-	ConvertImage(w, h, pixels, transparent);
+	if (pixels)
```
Comment (carstene1ns) [inline src/bitmap.cpp]: This is not necessary anymore, as you `return` before, in case of error.

File: src/bitmap.cpp [hunk 2]
```diff
@@ -559,17 +596,32 @@ Bitmap::Bitmap(const uint8_t* data, unsigned bytes, bool transparent, uint32_t f
...
+		Output::Warning("Unsupported image");
+
+	if (!img_okay) {
+		return;
+	}
 
 	Init(w, h, (void *) NULL);
-	ConvertImage(w, h, pixels, transparent);
+	if (pixels)
+		ConvertImage(w, h, pixels, transparent);
```
Comment (carstene1ns) [inline src/bitmap.cpp]: same.

**Comment by carstene1ns** [comment]
Looks good so far.
What about using the embedded System image (as mentioned in #856), does this work without changing anything (Cache)?

**Comment by Ghabry** [comment]
screenshot_0

Yeah it works I also tested invalid BMP and invalid XYZ codepaths yesterday, they behave the same.

One minor problem here is that the message is slightly misleading because it says "not found". The cache can't differentiate between not found and invalid because the return value is the same (empty image :/).  Hm but maybe I can fix this one by moving the warning to another location.

And the Output appears twice this is because our cache is a bit flush-friendly. I have plans to improve the cache but not in this PR :)

**Comment by Ghabry** [comment]
Thats better (deleted the title to show a difference)


screenshot_1

File: src/cache.cpp [hunk 3]
```diff
@@ -53,11 +53,18 @@ namespace {
...
 
+			BitmapRef bmp = BitmapRef();
+
 			if (path.empty()) {
-				return BitmapRef();
+				Output::Warning("Image not found: %s/%s", folder_name.c_str(), filename.c_str());
+			} else {
+				bmp = Bitmap::Create(path, transparent, flags);
+				if (!bmp) {
+					Output::Warning("Image invalid: %s/%s", folder_name.c_str(), filename.c_str());
```
Comment (fdelapena) [inline src/cache.cpp]: Nitpicky comment: maybe "Invalid image" is better.

Comment (Ghabry) [inline src/cache.cpp]: at least thats proper grammar

File: src/cache.cpp [hunk 4]
```diff
@@ -53,11 +53,18 @@ namespace {
...
+				}
 			}
-
-			return (cache[key] = Bitmap::Create(path, transparent, flags)).lock();
+
```
Comment (carstene1ns) [inline src/cache.cpp]: :o

File: src/bitmap.cpp [hunk 5]
```diff
@@ -947,7 +947,7 @@ void Bitmap::FillRect(Rect const& dst_rect, const Color &color) {
 void Bitmap::Clear() {
 	memset(pixels(), '\0', height() * pitch());
 
-        RefreshCallback();
+	RefreshCallback();
```
Comment (Ghabry) [inline src/bitmap.cpp]: before somebody complains here: This replaces 9 (!) spaces with 2 tabs :dancer:

**Comment by Ghabry** [comment]
Was there anything left here?

**Pull request merged**